### PR TITLE
Azure SQL Workspace Service

### LINKED ISSUE #3969: Azure SQL Workspace Service
A workspace service containing Azure SQL, following a similar pattern to the MySQL workspace service.

### PR DESCRIPTION
**Pull request by jonnyry**
# Changes

New Azure SQL workspace service, based on existing MySQL workspace

- New azuresql workspace service under `/templates/workspace-services/azuresql`
- TRE Core changes:
  - Add Azure SQL privatelink DNS zone to core terraform
- Base workspace changes:
  - Add Azure SQL privatelink DNS zone virtual network link to base workspace
- Documentation
- Add the workspace service template to `/.github/workflows/deploy_tre_reusable.yaml`

# Resolves #3969

## What is being addressed

Adds an Azure SQL workspace service following the same pattern as the My SQL workspace service.

## Follow ups

**A couple of follow up changes are required after this PR:**

~~1. The following repo is referenced in several places in the code https://github.com/microsoft/terraform-azurerm-environment-configuration.git.  Previously this was suffixed with ?ref=x.x.x, however I am waiting on a contributor to create a release so I can re-instate the reference.  This has been marked with the following comment in the code:  `# add specific ref once release 0.5.0 has been created`~~  (now completed - see commit dated 18-Jul)

2. Workspaces derived from base (i.e. unrestricted and airlock-import-review) both pull in the base workspace code from GitHub again using a ref.  Once a release has been created incorporating these changes, this ref needs updating in both workspace templates in the `Dockerfile.tmpl` file, variable `AZURE_TRE_VERSION`

### CODE REVIEW

**Comment by jonnyry** [comment]
Hi @marrobi @tim-allen-ck possible to get a review/merge on this one?  Thanks

**Comment by marrobi** [comment]
> Hi @marrobi @tim-allen-ck possible to get a review/merge on this one? Thanks

Tim focused on getting the release out last week. Will will make sure we test this out in the next week or so. Is it not being merged blocking you in any way?

**Comment by jonnyry** [comment]
> > Hi @marrobi @tim-allen-ck possible to get a review/merge on this one? Thanks
> 
> Tim focused on getting the release out last week. Will will make sure we test this out in the next week or so. Is it not being merged blocking you in any way?

Not strictly, as I can deploy from a branch with the changes in.  But preferable to have integrated into main.

**Comment by marrobi** [comment]
> > > Hi @marrobi @tim-allen-ck possible to get a review/merge on this one? Thanks
> > 
> > 
> > Tim focused on getting the release out last week. Will will make sure we test this out in the next week or so. Is it not being merged blocking you in any way?
> 
> Not strictly, as I can deploy from a branch with the changes in. But preferable to have integrated into main.

Ok, you can also add the bundle into your copy of the deployment repo.

**Comment by tim-allen-ck** [comment]
> Hi @marrobi @tim-allen-ck possible to get a review/merge on this one?  Thanks

I've left some comments on the code to look at.

**Comment by jonnyry** [comment]
> > Hi @marrobi @tim-allen-ck possible to get a review/merge on this one?  Thanks
> 
> I've left some comments on the code to look at.

Hi @tim-allen-ck Thanks, though I can't see any?  I might be looking in the wrong place.

**Comment by jonnyry** [comment]
@tim-allen-ck hey tim - I can't see the review comments you mentioned yesterday?  happy to demo this one if you like - hopefully more succesfully than yesterday :-D

**Comment by tim-allen-ck** [comment]
> @tim-allen-ck hey tim - I can't see the review comments you mentioned yesterday? happy to demo this one if you like - hopefully more succesfully than yesterday :-D

Hey, was just wanting you to use the most recent tf providers and an optional password field.

**Comment by jonnyry** [comment]
> Hey, was just wanting you to use the most recent tf providers and an optional password field.

I've updated the terraform version & provider versions.

Re the optional password field - could I suggest that we open a separate ticket for that which covers both MySQL & AzureSQL (and potentially looks at other auth options)?  I've gone with a straight conversion of the existing MySQL template to Azure SQL and not wanting to add new features in this PR.

**Review by tim-allen-ck — APPROVED** [review]
LGTM

File: templates/workspace_services/azuresql/terraform/main.tf [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# Azure Provider source and version being used
+terraform {
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = "3.18.0"
```
[the remark this answers is not in the document]
Comment (jonnyry) [reply templates/workspace_services/azuresql/terraform/main.tf]: Now updated 998070eb311861c196172964f0974b7f7c80ab1b

File: templates/workspace_services/azuresql/template_schema.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      "description": "Max storage allowed for a database (GB)",
+      "default": 5
+    },
+    "db_name": {
+      "$id": "#/properties/db_name",
+      "type": "string",
+      "title": "Database name",
+      "description": "Database name",
+      "default": "tredb"
+    }
```
[the remark this answers is not in the document]
Comment (jonnyry) [reply templates/workspace_services/azuresql/template_schema.json:37]: Re the optional password field - could I suggest that we open a separate ticket for that which covers both MySQL & AzureSQL (and potentially looks at other auth options)? I've gone with a straight conversion of the existing MySQL template to Azure SQL and not wanting to add new features in this PR.

**Review by marrobi — APPROVED** [review]
LGTM. As long as green tested can merge.

File: templates/workspace_services/azuresql/template_schema.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonnyry) [reply templates/workspace_services/azuresql/template_schema.json:37]: Hi @tim-allen-ck OK for this to be resolved?

**Pull request merged**